rules_haskell_tests/tests/RunTests.hs

```diff
@@ -165,6 +164,58 @@ bazel args = Process.proc "bazel" args
 bazelQuery :: String -> SpecM a [String]
 bazelQuery q = lines <$> runIO (Process.readProcess "bazel" ["query", q] "")
 
+-- | Shutdown Bazel
+shutdownBazel :: String -> IO ()
+shutdownBazel path = do
+  -- Related to https://github.com/tweag/rules_haskell/issues/2089
+  -- We experience intermittent "Exit Code: ExitFailure (-9)" errors. Shutdown 
+  -- Bazel when done executing tests for the workspace.
+  assertSuccess (bazel ["shutdown"]) { Process.cwd = Just path }
+  pure ()
+
+buildAndTest :: HasCallStack => String -> SpecWith ()
+buildAndTest path = describe path $ afterAll_ (shutdownBazel path) $ do
+  it "bazel build" $ do
+    assertSuccess $ (bazel ["build", "//..."]) { Process.cwd = Just path }
+  it "bazel test" $ do
+    assertSuccess $ (bazel ["test", "//..."]) { Process.cwd = Just path }
+
+-- * Print Memory Hooks
+
+-- | Print memory information before and after each test
+-- Only perform the hook if RHT_PRINT_MEMORY is "true".
+printMemoryHook :: IO () -> IO ()
+printMemoryHook action = do
+  rhtPrintMem <- lookupEnv "RHT_PRINT_MEMORY"
+  case rhtPrintMem of
+    Just "true" -> bracket_
+                     (printMemory "=== BEFORE ===")
+                     (printMemory "=== AFTER ===")
+                     action
+    _ -> action
+
+topPath :: String
+topPath = "/usr/bin/top"
+
+-- | Print information about the current memory state to debug intermittent failures
+-- Related to https://github.com/tweag/rules_haskell/issues/2089
+printMemory :: String -> IO ()
+printMemory msg = do
+  -- Do not attempt to run top, if it does not exist.
+  (exitCode, _, _) <- Process.readProcessWithExitCode "test" [topPath] ""
```

Comment:
Does this really check if the file exists? Don't you need to call it like `test -f path-to-file`?

You could also use `System.Directory.doesFileExist` instead.

Reply:
It doesn't. 🙂 I forgot the `-e`.  I will try using the suggested API.